### PR TITLE
Translation action for Hammer plugins

### PR DESCRIPTION
**Pull request by stejskalleos**
Action for pulling translations and updating strings.

**Workflow overview**
* Pull Hammer 
* Pull plugin
* Install and setup the Transifex
* Run the `tx:update`
* (check if something has been updated)
* Push the changes and create PR

**Requirements**
* tx_token: "${{ secrets.TX_TOKEN }}" Transifex API token
* gh_token: "${{ secrets.GITHUB_TOKEN }}" with permission to create a PR

**How to test**
Set the `TX_TOKEN` and update permissions for the `GITHUB_TOKEN`.
Then use `stejskalleos/foreman_actions/` repo for the action
```
name: Update translations

on:
  workflow_dispatch:

jobs:
  tx-update:
    name: Update translations
    uses: stejskalleos/foreman_actions/.github/workflows/tx-hammer-plugins.yml@hammer-tx
    with:
      plugin_name: "hammer-cli-foreman-google"
    secrets:
      tx_token: "${{ secrets.TX_TOKEN }}"
      gh_token: "${{ secrets.GITHUB_TOKEN }}"
```

### CODE REVIEW

File: .github/workflows/tx-hammer-plugins.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+name: Translations
+
+on:
+  workflow_call:
+    inputs:
+      plugin_name:
+        description: 'Name of the Hammer plugin. (Not the repository name)'
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml]: Is this the gem name?

File: .github/workflows/tx-hammer-plugins.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          sudo mv tx /usr/local/bin/
+          sudo chown -R runner:runner /usr/local/bin
+
+      - name: Configure plugin
+        run: |
+          echo "gemspec path: '../hammer-cli', name: 'hammer_cli'" > Gemfile.local.rb
+          mkdir -p ./.tx
+          echo "[main]" > ./.tx/config
+          echo "host = https://rest.api.transifex.com" >> ./.tx/config
+          echo "token = ${{ secrets.tx_token  }}" >> ./.tx/config
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml]: I'd consider the multi line statement
```suggestion
          cat > ./.tx/config <<EOF
          [main]
          host = https://rest.api.transifex.com
          token = ${{ secrets.tx_token  }}
          EOF
```

**Review by ekohl — COMMENTED** [review]
I like the idea and think we should do something similar for Foreman plugins.

I think with hammer it is possible to clone a plugin and run the action from there. You simply make sure it pulls Hammer from git using the `Gemfile`. Would that simplify the action?

File: .github/workflows/tx-hammer-plugins.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        run: |
+          if ! [ -n "$(git diff ${{ github.ref_name }})" ]; then
+            echo '###'
+            echo '### No changes.'
+            echo '### Exiting ...'
+            echo '###'
+            exit 1
+          fi
+
+      - name: Push changes and create a pull request
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml]: In packaging we have this bit: https://github.com/theforeman/foreman-packaging/blob/55448ce7ffbee75b9edd16f54b22030580fada5d/.github/workflows/bump_packages.yml#L64-L71

We know that one works well and also deals with an already existing branch + PR.

**Review by ekohl — COMMENTED** [review]
Forgot to submit one comment.

File: .github/workflows/tx-hammer-plugins.yml (hunk 3 above)
Comment (stejskalleos) [reply .github/workflows/tx-hammer-plugins.yml]: Yeah, I can use that, I wasn't sure what's the policy for using actions from the "outside".

File: .github/workflows/tx-hammer-plugins.yml (hunk 1 above)
Comment (stejskalleos) [reply .github/workflows/tx-hammer-plugins.yml]: Yop. I will rename it

**Comment by stejskalleos** [comment]
> I like the idea and think we should do something similar for Foreman plugins.

That's the plan. First I would like to use it just for the Hammer plugins, spread it across multiple repositories, and after we catch up on all the bugs & enhancements I will create action for the Foreman plugins.


> I think with Hammer it is possible to clone a plugin and run the action from there. You simply make sure it pulls Hammer from git using the Gemfile. Would that simplify the action?

Maybe a little, I still have to go to the gem folder and to the git stuff there.

**Comment by stejskalleos** [comment]
@ekohl updated, did some cleanup and use gh action for the PR stuff

File: .github/workflows/tx-hammer-plugins.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          host = https://rest.api.transifex.com
+          token = ${{ secrets.tx_token  }}
+
+          [o:foreman:p:foreman:r:${{ inputs.gem_name }}]
+          file_filter = locale/<lang>/${{ inputs.gem_name }}.edit.po
+          source_file = locale/${{ inputs.gem_name }}.pot
+          source_lang = en
+          type = PO
+          resource_name = ${{ inputs.gem_name }}
+          EOF
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:63]: I'd prefer to keep the transifex config in the repo itself so you can also run it locally. You already set `TX_TOKEN` as an env var so that's already done.
```suggestion
```

File: .github/workflows/tx-hammer-plugins.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: "Install Transifex"
+        run: |
+          curl -o- https://raw.githubusercontent.com/transifex/cli/master/install.sh | bash
+          sudo mv tx /usr/local/bin/
+          sudo chown -R runner:runner /usr/local/bin
+
+      - name: Configure plugin
+        run: |
+          echo "gemspec path: '../hammer-cli', name: 'hammer_cli'" > Gemfile.local.rb
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:50]: I don't see the point of this. Hammer doesn't declare any development dependencies so I'd think this is equivalent:
```suggestion
          echo "gem 'hammer_cli', github: 'theforeman/hammer-cli'" > Gemfile.local.rb
```

Then you don't need to clone it separately.

A further question: how many Gemfiles actually load `Gemfile.local.rb`?

File: .github/workflows/tx-hammer-plugins.yml [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      - name: Pull translations
+        env:
+          TX_TOKEN: ${{ secrets.tx_token  }}
+        run: |
+          bundle exec rake tx:update
+
+      - name: Cleanup before the PR
+        run: |
+          git checkout -f
+          rm -rf vendor
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:86]: I'd like to avoid this and instead make sure `vendor` is in `.gitignore`

File: .github/workflows/tx-hammer-plugins.yml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: Pull translations
+        env:
+          TX_TOKEN: ${{ secrets.tx_token  }}
+        run: |
+          bundle exec rake tx:update
+
+      - name: Cleanup before the PR
+        run: |
+          git checkout -f
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:85]: Why is this needed?

File: .github/workflows/tx-hammer-plugins.yml [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+          working-directory: ${{ inputs.gem_name }}
+
+      - name: Set up Git
+        run: |
+          git config user.name "Foreman Translation action"
+          git config user.email "noreply@foreman.example.com"
+
+      - name: Pull translations
+        env:
+          TX_TOKEN: ${{ secrets.tx_token  }}
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:79]:
```suggestion
          TX_TOKEN: ${{ secrets.tx_token }}
```

File: .github/workflows/tx-hammer-plugins.yml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        uses: actions/checkout@v4
+        with:
+          repository: theforeman/hammer-cli
+          ref: master
+          path: hammer-cli
+
+      - name: Get ${{ github.repository }}
+        uses: actions/checkout@v4
+        with:
+          path: ${{ inputs.gem_name  }}
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:36]:
```suggestion
          path: ${{ inputs.gem_name }}
```

File: .github/workflows/tx-hammer-plugins.yml [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      gh_token:
+        description: 'GitHub token with permissions to create a pull request'
+        required: true
+
+jobs:
+  tx-update:
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: ${{ inputs.gem_name  }}
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:23]:
```suggestion
        working-directory: ${{ inputs.gem_name }}
```

File: .github/workflows/tx-hammer-plugins.yml [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: Install Transifex dependencies
+        run: |
+          sudo apt-get -y install gettext
+
+      - name: "Install Transifex"
+        run: |
+          curl -o- https://raw.githubusercontent.com/transifex/cli/master/install.sh | bash
+          sudo mv tx /usr/local/bin/
+          sudo chown -R runner:runner /usr/local/bin
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:46]: I don't see a reason to chown it. Normally things in `/usr/local/bin` are `root:root` mode 0755 which works fine. In this particular case you already download it as `runner` and `mv` maintains that so if anything, I think this is redundant:
```suggestion
```

File: .github/workflows/tx-hammer-plugins.yml [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        description: 'Name of the Gem'
+        required: true
+        type: 'string'
+    secrets:
+      tx_token:
+        description: 'API token for the Transifex.'
+        required: true
+      gh_token:
+        description: 'GitHub token with permissions to create a pull request'
+        required: true
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:16]: I'd think this isn't needed. GitHub already has the right permissions if it runs as an action.

File: README.md [hunk 13]
```diff
@@ -100,3 +100,21 @@ jobs:
     secrets:
       api_key: ${{ secrets.RUBYGEM_API_KEY }}
 ```
+
+## Update translations - Hammer plugins
```
Comment (ekohl) [inline README.md:104]:
```suggestion
## Update translations - Hammer plugins

```

File: .github/workflows/tx-hammer-plugins.yml [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: "2.7"
+          bundler-cache: true
+          working-directory: ${{ inputs.gem_name }}
+
+      - name: Set up Git
+        run: |
+          git config user.name "Foreman Translation action"
+          git config user.email "noreply@foreman.example.com"
```
Comment (ekohl) [inline .github/workflows/tx-hammer-plugins.yml:75]: I wouldn't use `.example.com` in a real situation. This will end up in git logs. In our other action we do this: https://github.com/theforeman/foreman-packaging/blob/99e5428c8ef99e38bd6b015334589c6e2ade2a66/.github/workflows/bump_packages.yml#L8-L10

Note it also uses env vars instead of running `git config`.

**Review by ekohl — CHANGES_REQUESTED** [review]
GitHub Actions should be as simple as possible and trivial to replicate locally or in a different CI system. Logic in it should be avoided as much as possible.

If you let bundler do the cloning of `hammer-cli` then you don't need to set up a separate cloning, use different working directories and require the `gem_name` input.
